kobo/apps/subsequences/utils/deprecation.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import Optional
+
+import jsonschema
+
+from kpi.fields import WritableJSONField
+from ..advanced_features_params_schema import (
+    ADVANCED_FEATURES_PARAMS_SCHEMA,
+)
+
+
+def jsonschema_validate(asset: 'Asset'):
+    try:
+        jsonschema.validate(
+            instance=asset.advanced_features,
+            schema=ADVANCED_FEATURES_PARAMS_SCHEMA,
+        )
+    except jsonschema.exceptions.ValidationError as e:
+        if "'qpath' was unexpected" not in str(e):
+            raise
+
+        qual_survey_orig = asset.advanced_features['qual']['qual_survey']
+        qual_survey_iter = deepcopy(qual_survey_orig)
+        for idx, qual_q in enumerate(qual_survey_iter):
+            qpath = qual_survey_orig[idx]['qpath']
+            xpath = qpath_to_xpath(qpath, asset)
+            del qual_survey_orig[idx]['qpath']
+            qual_survey_orig[idx]['xpath'] = xpath
+
+        jsonschema.validate(
+            instance=asset.advanced_features,
+            schema=ADVANCED_FEATURES_PARAMS_SCHEMA,
+        )
+
+
+def qpath_to_xpath(qpath: str, asset: 'Asset') -> str:
+    """
+    We have abandoned `qpath` attribute in favor of `xpath`.
+    Existing projects may still use it though.
+    We need to find the equivalent `xpath`
+    """
+    for row in asset.content['survey']:
+        if '$qpath' in row and '$xpath' in row and row['$qpath'] == qpath:
+            return row['$xpath']
+
+    # Could not find it from the survey, let's try to detect it automatically
+    xpaths = asset.get_attachment_xpaths(deployed=True)
+    for xpath in xpaths:
+        dashed_xpath = xpath.replace('/', '-')
+        if dashed_xpath == qpath:
+            return xpath
+
+    raise KeyError(f'xpath for {qpath} not found')
+
+
+def sanitize_known_columns(asset: 'Asset'):
+    for idx, known_column in enumerate(asset.known_cols):
+        xpath, *rest = known_column.split(':')
+        # Old `qpath` should not contain "/", but could contain "-".
+        # If the question does not belong to a group but does contain "-",
+        # it will enter this condition - which is not a problem except extra
+        # CPU usage for nothing.
+        if '-' in xpath and '/' not in xpath:
+            xpath = qpath_to_xpath(xpath, asset)
+            rest.insert(0, xpath)
+            asset.known_cols[idx] = ':'.join(rest)
+
+    # TODO Should we save asset.known_cols if it has changed?
```

Comment:
Same as below, pass the dictionary and return the new dictionary, instead of passing and mutating the asset instance

Reply:
Unfortunately, we need the asset object in `qpath_to_xpath()`. It would be a little bit redundant to pass `asset.known_cols` and `asset` to `sanitize_known_columns()`. 

What do you think about something like that? 

```python
def get_sanitize_known_columns(asset: 'Asset') -> dict | None:
    known_cols = deepcopy(asset.known_cols)
    for idx, known_column in enumerate(known_cols):
        ...
    return knows_cols
```

Reply:
yeah, it works